### DISCUSSION TITLE
Proper way and event for use of IsFieldChanged

### DISCUSSION THREAD

**Discussion by LSNicholls** [serenity-is/Serenity#6596] [Help] [answered]
Hi folks,

I wanted to update a couple of fields represent who uploaded a document and when, and noted advice in #5831, so I thought **BeforeSave** would be a good place to do something similar. 

In original tests it worked fine, but in this case I *only* want to change the values when the actual uploaded document has changed, not when other values in the row, characterizing the upload, had changed.

I noted advice in #5583 and thought **IsFieldChanged**  would give me what I wanted.  But in BeforeSave, I think maybe it's too late already to use IsFieldChanged, because adding this test never gave me a positive result.

I found that I could do what I wanted in PerformAuditing, but I I still couldn't get IsFieldChanged to work.  Right now my code is shown as below.  ( In this code, CodedFileName represents the pathed uploaded document as provided by the FileUploadEditor.)

This code is working but I have a feeling I'm probably not doing it the best way.  Does anybody have advice on where and when IsFieldChanged would have resolved this in a more elegant way?

Thank you!

```
       protected override void PerformAuditing()
        {
            base.PerformAuditing();
                    
            if ((!(Row.CodedFileName == null)) && ((this.Old.CodedFileName == null) || 
                 (this.Old.CodedFileName != Row.CodedFileName) ))
            {
                Row.Uploader = Convert.ToInt32(User.GetIdentifier());
                Row.Uploaded = DateTime.Now;
            }
        }
```

**Comment by volkanceylan** [comment, marked as answer]
IsFieldChanged is something usually meant for desktop apps, and requires calling BeginEdit to work. Think of WinForms grids like DevExpress one where you enter edit mode while editing an entity and raise a set of IPropertyChanged events etc.

PerformAuditing is not the right place to change an entity if you want these changes to be saved to the database. As it is executed after save, just like AfterSave, even after the AfterSave :) It is used to perform auditing after the operation, e.g. AuditLog, CaptureLog etc.

So, BeforeSave is better, but after calling base.BeforeSave as other behaviors are executed there in the save handler. To understand if a field is changed and it is set to a non null value, something like `Row.CodedFileName != null && Row.CodedFileName != Old?.CodedFileName` should suffice. Old can be null in Create operations.

**Comment by LSNicholls** [comment]
>IsFieldChanged is something usually meant for desktop apps, and requires calling BeginEdit to work. Think of WinForms grids like DevExpress one where you enter edit mode while editing an entity and raise a set of IPropertyChanged events etc.

I see thanks.
